CitiesSkylinesMods/TMPE
# TMPE support in editor

[TrafficManager.zip]( https://ci.appveyor.com/api/projects/krzychu124/tmpe/artifacts/TMPE.zip?branch=major-feature-asset-editor) from appvoyer.
### Review Process:
this is a major feature branch. PR to major feature branch need only one approval and PR from major feature branch to master need two approvals (at least one developer. the second one could be chamelon)
In previous PRs all the code for this branch has been reviewed together with some minor testing. 
This PR is about full-proof testing.
### Notes:
- `appveyor` has latest [TrafficManager.zip]( https://ci.appveyor.com/api/projects/krzychu124/tmpe/artifacts/TMPE.zip?branch=major-feature-asset-editor) 
- Vehicle restriction is another PR #966. if you wish to test Vehicle restriction then merge that branch. but it should not be necessary because you can test lane serialization with lane arrows.
- The timed traffic light is not supported see #967 
- this code is compatible with Smart intersection mod (because I moved it to harmony).

### Some test scenarios that came to my mind:
- use the intersection editor to create an intersection with TMPE rules.
   - place that intersection in-game and check if TMPE rules persist.
       - try both with and without smart intersection mod.
   - use `Editor->asset editor->NEW  ` to load the intersection from loaded prefab collection and check if traffic rules persist.
   - disable the intersection from content manager then use `Editor->asset editor->LOAD` tp load the intersection from file and check if traffic rules persist. 
- load a map in the map editor. create some TMPE rules and save it
   - play the map and check if TMPE rules persist.
   - load the map and check if TMPE rules persist.
